x/ccv/provider/keeper/distribution.go

```diff
@@ -67,142 +66,265 @@ func (k Keeper) GetAllConsumerRewardDenoms(ctx sdk.Context) (consumerRewardDenom
+		k.Logger(ctx).Error(
+			"fail to allocate ICS rewards to community pool",
+			"consumerId", consumerId,
+			"chainId", chainId,
+			"error", err.Error(),
 		)
+		return types.ConsumerRewardsAllocation{}, err
+	}
+
+	// set consumer allocations to the remaining rewards decimals
+	alloc.Rewards = validatorsRewardsChange.Add(remainingChanges...)
+
+	k.Logger(ctx).Info(
+		"distributed ICS rewards successfully",
+		"consumerId", consumerId,
+		"chainId", chainId,
+		"total-rewards", consumerRewards.String(),
+		"sent-to-distribution", validatorsRewardsTrunc.String(),
+		"sent-to-CP", remainingRewards.String(),
+	)
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeDistributedRewards,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
+			sdk.NewAttribute(types.AttributeConsumerId, consumerId),
+			sdk.NewAttribute(types.AttributeConsumerChainId, chainId),
+			sdk.NewAttribute(types.AttributeRewardTotal, consumerRewards.String()),
+			sdk.NewAttribute(types.AttributeRewardDistributed, validatorsRewardsTrunc.String()),
+			sdk.NewAttribute(types.AttributeRewardCommunityPool, remainingRewards.String()),
+		),
+	)
+	return alloc, nil
+}
+
+// AllocateTokens performs rewards distribution to the community pool and validators
+// based on the Partial Set Security distribution specification.
+func (k Keeper) AllocateTokens(ctx sdk.Context) {
+	// return if there is no coins in the consumer rewards pool
+	if k.GetConsumerRewardsPool(ctx).IsZero() {
+		return
+	}
+
+	// Iterate over all launched consumer chains.
+	// To avoid large iterations over all the consumer IDs, iterate only over
+	// chains with an IBC client created.
+	allConsumerRewardDenoms := k.GetAllConsumerRewardDenoms(ctx) // corresponds to allowlisted denoms that were allowlisted through governance
+	for _, consumerId := range k.GetAllConsumersWithIBCClients(ctx) {
+		// also consider this chain's allowlisted reward denoms
+		consumerAllowlistedRewardDenoms, err := k.GetAllowlistedRewardDenoms(ctx, consumerId)
+		if err != nil {
+			k.Logger(ctx).Error(
+				"fail to retrieve the allowlisted reward denoms for consumer chain",
+				"consumer id", consumerId,
+				"error", err.Error())
+		}
+
+		allAllowlistedDenoms := append(allConsumerRewardDenoms, consumerAllowlistedRewardDenoms...)
+		for _, denom := range allAllowlistedDenoms {
+			// use a cached context to verify that the call to `AllocateConsumerRewards` is atomic, and hence
+			// all transfers in `AllocateConsumerRewards` happen all together or not at all.
+			cachedCtx, writeCache := ctx.CacheContext()
+			consumerRewards, err := k.GetConsumerRewardsAllocationByDenom(cachedCtx, consumerId, denom)
+			if err != nil {
+				k.Logger(ctx).Error(
+					"failed to get the consumer rewards allocation for this denom",
+					"consumer id", consumerId,
+					"denom", denom,
+					"error", err.Error(),
+				)
+			}
+			allocatedRewards, err := k.AllocateConsumerRewards(cachedCtx, consumerId, consumerRewards)
```

Comment:
nit: could you please rename `allocatedRewards` to stg like `remainingRewardDec`?